File: transports/ocaml/lib/rpc.ml
(* See rpc.mli for the contract. Ported from transports/go/rpc.go. *)

open Conventions

type request = {
  service : string;
  op : string;
  id : int64 option;
  payload : bytes;
  auth : string option;
}

let new_request service op payload =
  { service; op; id = None; payload; auth = None }

let request_with_id r id = { r with id = Some id }
let request_with_auth r auth = { r with auth = Some auth }

let encode_request (r : request) =
  let entries =
    [
      (Cbor.Text "v", Cbor.Uint version);
      (Cbor.Text "service", Cbor.Text r.service);
      (Cbor.Text "op", Cbor.Text r.op);
      (Cbor.Text "payload", tag24 r.payload);
    ]
  in
  let entries =
    match r.id with
    | Some id -> entries @ [ (Cbor.Text "id", Cbor.Uint id) ]
    | None -> entries
  in
  let entries =
    match r.auth with
    | Some a -> entries @ [ (Cbor.Text "auth", Cbor.Text a) ]
    | None -> entries
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_request b =
  let* v = decode_envelope b in
  let* ver = get_uint v "v" in
  let* () = check_version ver in
  match map_get v "payload" with
  | None -> Error (malformed "missing 'payload'")
  | Some p ->
      let* payload = untag24 p in
      let* service = get_text v "service" in
      let* op = get_text v "op" in
      Ok
        {
          service;
          op;
          id = get_uint_opt v "id";
          payload;
          auth = get_text_opt v "auth";
        }

type response = {
  id : int64 option;
  status : Conventions.status;
  variant : string option;
  error : string option;
  payload : bytes;
}

let new_response_ok variant payload =
  {
    id = None;
    status = status_ok;
    variant = Some variant;
    error = None;
    payload;
  }

let new_response_transport_error status message =
  {
    id = None;
    status;
    variant = None;
    error = Some message;
    payload = Bytes.create 0;
  }

let response_with_id r id = { r with id }

let encode_response (r : response) =
  let entries =
    [
      (Cbor.Text "v", Cbor.Uint version);
      (Cbor.Text "status", Cbor.Int (status_code r.status));
      (Cbor.Text "payload", tag24 r.payload);
    ]
  in
  let entries =
    match r.id with
    | Some id -> entries @ [ (Cbor.Text "id", Cbor.Uint id) ]
    | None -> entries
  in
  let entries =
    match r.variant with
    | Some v -> entries @ [ (Cbor.Text "variant", Cbor.Text v) ]
    | None -> entries
  in
  let entries =
    match r.error with
    | Some e -> entries @ [ (Cbor.Text "error", Cbor.Text e) ]
    | None -> entries
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_response b =
  let* v = decode_envelope b in
  let* ver = get_uint v "v" in
  let* () = check_version ver in
  (* payload is present but may wrap an empty byte string on a non-zero status. *)
  let* payload =
    match map_get v "payload" with
    | Some p -> untag24 p
    | None -> Ok (Bytes.create 0)
  in
  let* status = get_int v "status" in
  Ok
    {
      id = get_uint_opt v "id";
      status = status_from_code status;
      variant = get_text_opt v "variant";
      error = get_text_opt v "error";
      payload;
    }

let as_transport_error (r : response) =
  if status_is_ok r.status then None
  else
    let message = match r.error with Some m -> m | None -> "" in
    Some
      (Status_error
         { status = status_name r.status; code = status_code r.status; message })

type push = { service : string; event : string; payload : bytes }

let new_push service event payload = { service; event; payload }

let encode_push (p : push) =
  let entries =
    [
      (Cbor.Text "v", Cbor.Uint version);
      (Cbor.Text "service", Cbor.Text p.service);
      (Cbor.Text "event", Cbor.Text p.event);
      (Cbor.Text "payload", tag24 p.payload);
    ]
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_push b =
  let* v = decode_envelope b in
  let* ver = get_uint v "v" in
  let* () = check_version ver in
  match map_get v "payload" with
  | None -> Error (malformed "missing 'payload'")
  | Some p ->
      let* payload = untag24 p in
      let* service = get_text v "service" in
      let* event = get_text v "event" in
      Ok { service; event; payload }

type handler_outcome =
  | Reply of { variant : string; payload : bytes }
  | Transport_status of { status : Conventions.status; message : string }

let reply variant payload = Reply { variant; payload }
let transport status message = Transport_status { status; message }

type client = {
  carrier : Carrier.frame_carrier;
  mutable next_id : int64;
  multiplexed : bool;
}

let new_client carrier multiplexed = { carrier; next_id = 1L; multiplexed }

let call (client : client) ~service ~op ~payload ?auth () =
  let req = { (new_request service op payload) with auth } in
  let req =
    if client.multiplexed then (
      let id = client.next_id in
      client.next_id <- Int64.add id 1L;
      { req with id = Some id })
    else req
  in
  let frame = encode_request req in
  let* () = client.carrier.Carrier.send_frame frame in
  let* incoming = client.carrier.Carrier.recv_frame () in
  match incoming with
  | None -> Error (Carrier "connection closed before response")
  | Some inb -> (
      let* resp = decode_response inb in
      match as_transport_error resp with Some e -> Error e | None -> Ok resp)

type server = { carrier : Carrier.frame_carrier }

let new_server carrier = { carrier }

let serve_one (server : server) handler =
  let* frame = server.carrier.Carrier.recv_frame () in
  match frame with
  | None -> Ok false
  | Some f ->
      let resp =
        match decode_request f with
        | Ok req -> (
            let id = req.id in
            match handler req with
            | Reply { variant; payload } ->
                response_with_id (new_response_ok variant payload) id
            | Transport_status { status; message } ->
                response_with_id
                  (new_response_transport_error status message)
                  id)
        | Error e ->
            new_response_transport_error status_malformed_envelope
              (error_message e)
      in
      let out = encode_response resp in
      let* () = server.carrier.Carrier.send_frame out in
      Ok true

File: transports/php/src/Rpc.php
<?php

namespace Csilgen\Transport;

/**
 * CSIL-RPC transport -- request/response/push envelopes -- see csil-rpc-transport.md.
 *
 * Payloads are opaque CBOR(type) bytes; on the wire they ride inside tag 24. Map
 * keys are deterministically ordered by CBOR::encode, so the same logical envelope
 * always yields the same bytes as the Rust/Go reference libraries.
 */
final class Rpc
{
    /**
     * Encode a CsilRpcRequest (spec section 1.1): {v, service, op, ?id, payload, ?auth}.
     * `id` is required on multiplexed carriers, omitted on one-in-flight carriers.
     */
    public static function encodeRequest($service, $op, $payload, $id = null, $auth = null)
    {
        $map = array(
            'v' => Conventions::VERSION,
            'service' => $service,
            'op' => $op,
            'payload' => Conventions::tag24($payload),
        );
        if ($id !== null) {
            $map['id'] = $id;
        }
        if ($auth !== null) {
            $map['auth'] = $auth;
        }
        return CBOR::encode($map);
    }

    /**
     * @return array{service:string,op:string,id:?int,payload:string,auth:?string}
     */
    public static function decodeRequest($bytes)
    {
        $m = Conventions::decodeEnvelope($bytes);
        Conventions::checkVersion(Conventions::getUint($m, 'v'));
        if (!isset($m['payload'])) {
            throw new MalformedException("missing 'payload'");
        }
        return array(
            'service' => Conventions::getText($m, 'service'),
            'op' => Conventions::getText($m, 'op'),
            'id' => Conventions::getUintOpt($m, 'id'),
            'payload' => Conventions::untag24($m['payload']),
            'auth' => Conventions::getTextOpt($m, 'auth'),
        );
    }

    /**
     * Encode a CsilRpcResponse (spec section 1.2): {v, ?id, status, ?variant, ?error, payload}.
     * `payload` is empty when `status` is non-zero, but stays present as an empty
     * tag-24 byte string so success and failure share one envelope shape.
     */
    public static function encodeResponse($status, $payload, $id = null, $variant = null, $error = null)
    {
        $map = array(
            'v' => Conventions::VERSION,
            'status' => $status,
            'payload' => Conventions::tag24($payload),
        );
        if ($id !== null) {
            $map['id'] = $id;
        }
        if ($variant !== null) {
            $map['variant'] = $variant;
        }
        if ($error !== null) {
            $map['error'] = $error;
        }
        return CBOR::encode($map);
    }

    /**
     * @return array{id:?int,status:int,variant:?string,error:?string,payload:string}
     */
    public static function decodeResponse($bytes)
    {
        $m = Conventions::decodeEnvelope($bytes);
        Conventions::checkVersion(Conventions::getUint($m, 'v'));
        // A non-zero-status response may omit the payload entirely.
        $payload = isset($m['payload']) ? Conventions::untag24($m['payload']) : '';
        return array(
            'id' => Conventions::getUintOpt($m, 'id'),
            'status' => Conventions::getInt($m, 'status'),
            'variant' => Conventions::getTextOpt($m, 'variant'),
            'error' => Conventions::getTextOpt($m, 'error'),
            'payload' => $payload,
        );
    }

    /**
     * Return the decoded response unchanged when its status is ok; otherwise throw
     * StatusException. Callers use this after decodeResponse to surface transport
     * failures distinctly from application errors (which ride as a status-0 typed
     * variant inside the payload).
     */
    public static function checkStatus(array $response)
    {
        if ($response['status'] !== Conventions::STATUS_OK) {
            throw new StatusException($response['status'], $response['error']);
        }
        return $response;
    }

    /**
     * Encode a CsilRpcPush (spec section 1.3): {v, service, event, payload}.
     */
    public static function encodePush($service, $event, $payload)
    {
        return CBOR::encode(array(
            'v' => Conventions::VERSION,
            'service' => $service,
            'event' => $event,
            'payload' => Conventions::tag24($payload),
        ));
    }

    /**
     * @return array{service:string,event:string,payload:string}
     */
    public static function decodePush($bytes)
    {
        $m = Conventions::decodeEnvelope($bytes);
        Conventions::checkVersion(Conventions::getUint($m, 'v'));
        if (!isset($m['payload'])) {
            throw new MalformedException("missing 'payload'");
        }
        return array(
            'service' => Conventions::getText($m, 'service'),
            'event' => Conventions::getText($m, 'event'),
            'payload' => Conventions::untag24($m['payload']),
        );
    }
}
